platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/CountUpLatch.java

```diff
@@ -0,0 +1,159 @@
+     * the count. It is, however, safe for many threads to concurrently call methods that query the count.
+     */
+    public void increment() {
+        add(1);
+    }
+
+    /**
+     * Set the count to a higher value.
+     * <p>
+     * Methods that update the count are not mutually thread safe. Concurrent threads should never attempt to update
+     * the count. It is, however, safe for many threads to concurrently call methods that query the count.
+     *
+     * @param count the new count, must be greater than the current value
+     * @throws IllegalArgumentException if the new count is less than the current count (the state of this object is
+     *                                  messed up if this happens -- this is an unrecoverable error)
+     */
+    public void set(final long count) {
+        final long previous = currentCount.getAndSet(count);
+        if (count < previous) {
+            throw new IllegalArgumentException("Current count is " + previous + " but new count is " + count
+                    + ", cannot set count to a lower value");
+        }
+        phaser.arriveAndAwaitAdvance();
+    }
+
+    /**
+     * Increment the count by the given delta.
+     * <p>
+     * Methods that update the count are not mutually thread safe. Concurrent threads should never attempt to update
+     * the count. It is, however, safe for many threads to concurrently call methods that query the count.
+     *
+     * @param delta the amount to increment the count by
+     * @throws IllegalArgumentException if the delta is negative
+     */
+    public void add(final long delta) {
+        if (delta < 0) {
+            throw new IllegalArgumentException("Cannot add a negative delta (" + delta + ")");
+        }
+        currentCount.addAndGet(delta);
+        phaser.arriveAndAwaitAdvance();
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count the count to wait for
+     * @throws InterruptedException if interrupted while waiting
+     */
+    public void await(final long count) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return;
+        }
+
+        int phase = phaser.getPhase();
+        while (currentCount.get() < count) {
+            phase = phaser.awaitAdvance(phase);
+        }
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count      the count to wait for
+     * @param timeToWait the maximum time to wait
+     * @return true if the count reached the given value, false if the time to wait expired
+     * @throws InterruptedException if interrupted while waiting
+     */
+    public boolean await(final long count, final Duration timeToWait) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return true;
+        }
+
+        final Instant start = Instant.now();
```

Comment:
Suppose the counter is updated once per millisecond, and we set the timeout to one second. In that scenario, await would block indefinitely until the count reaches the desired value.

Reply:
The wait would exit at 1 second, with timeout exception, if it did not exit sooner.  It will not wait indefinitely if a timeout is set.
The timeout parameter is the maximum time to wait, it will wake when the condition is signaled regardless, and wake with timeout when the time expires if it has not awoken before that.
A thread may experience a spurious wake at any time, which is why you usually need to recheck conditions after a wait, to ensure the conditions are what you expect.